lib/css-converter.js

```diff
@@ -0,0 +1,257 @@
+import { CssSelectorParser } from 'css-selector-parser';
+import log from './logger';
+
+const CssConverter = {};
+
+const parser = new CssSelectorParser();
+parser.registerSelectorPseudos('has');
+parser.registerNestingOperators('>', '+', '~');
+parser.registerAttrEqualityMods('^', '$', '*', '~');
+parser.enableSubstitutes();
+
+const booleanAttrs = [
+  'checkable', 'checked', 'clickable', 'enabled', 'focusable',
+  'focused', 'long-clickable', 'scrollable', 'selected',
+];
+
+const numericAttrs = [
+  'index', 'instance',
+];
+
+const strAttrs = [
+  'description', 'resource-id', 'text', 'class-name', 'package-name'
+];
+
+const allAttrs = [
+  ...booleanAttrs,
+  ...numericAttrs,
+  ...strAttrs,
+];
+
+const attributeAliases = [
+  ['resource-id', ['id']],
+  ['description', [
+    'content-description', 'content-desc',
+    'desc', 'accessibility-id',
+  ]],
+  ['index', ['nth-child']],
+];
+
+/**
+ * Convert hyphen separated word to snake case
+ * @param {String} str
+ */
+function toSnakeCase (str) {
+  const tokens = str.split('-').map((str) => str.charAt(0).toUpperCase() + str.slice(1).toLowerCase());
+  let out = tokens.join('');
+  return out.charAt(0).toLowerCase() + out.slice(1);
+}
+
+/**
+ * Get the boolean from a CSS object. If empty, return true. If not true/false/empty, throw exception
+ *
+ * @param {*} css CSS object
+ */
+function assertGetBool (css) {
+  const val = css.value?.toLowerCase() || 'true';
+  if (['true', 'false'].includes(val)) {
+    return val;
+  }
+  log.errorAndThrow(`Could not parse '${css.name}=${css.value}'. '${css.name}' must be true, false or empty`);
+}
+
+/**
+ * Get the canonical form of a CSS attribute name
+ *
+ * Converts to lowercase and if an attribute name is an alias for something else, return
+ * what it is an alias for
+ *
+ * @param {*} css CSS object
+ */
+function assertGetAttrName (css) {
+  const attrName = css.name.toLowerCase();
+
+  // Check if it's supported and if it is, return it
+  if (allAttrs.includes(attrName)) {
+    return attrName.toLowerCase();
+  }
+
+  // If attrName is an alias for something else, return that
+  for (let [officialAttr, aliasAttrs] of attributeAliases) {
+    if (aliasAttrs.includes(attrName)) {
+      return officialAttr;
+    }
+  }
+  log.errorAndThrow(`'${attrName}' is not a valid attribute. ` +
+    `Supported attributes are '${allAttrs.join(', ')}'`);
+}
+
+/**
+ * Get a regex that matches a whole word. For the ~= CSS attribute selector.
+ *
+ * @param {string} word
+ */
+function getWordMatcherRegex (word) {
+  return `\\b(\\w*${escapeRegexLiterals(word)}\\w*)\\b`;
+}
+
+/**
+ * Add escapes to regex literals
+ *
+ * @param {*} str
+ */
+function escapeRegexLiterals (str) {
```

Comment:
lodash already has a helper to escape regex

Reply:
Found this: https://stackoverflow.com/questions/3561493/is-there-a-regexp-escape-function-in-javascript ... copied and pasted. I couldn't find lodash version of this.

Reply:
https://lodash.com/docs/4.17.15#escapeRegExp